IdentityCore/src/cache/mac/MSIDMacKeychainTokenCache.m

```diff
@@ -457,72 +481,22 @@ - (BOOL)saveToken:(MSIDCredentialCacheItem *)credential
     
-    if (savedStorageItem)
+    if (bucket)
     {
-        [storageItem mergeStorageItem:savedStorageItem];
-    }
-    
-    [storageItem storeCredential:credential forKey:(MSIDDefaultCredentialCacheKey *)key];
-    return [self saveStorageItem:storageItem key:key serializer:serializer context:context error:error];
-    
-    MSID_LOG_WITH_CTX_PII(MSIDLogLevelError, context, @"Failed to save stored credential for key %@.", MSID_PII_LOG_MASKABLE(key));
-    return NO;
-}
-
-- (BOOL)saveStorageItem:(MSIDMacCredentialStorageItem *)storageItem
-                    key:(MSIDCacheKey *)key
-             serializer:(id<MSIDCacheItemSerializing>)serializer
-                context:(id<MSIDRequestContext>)context
-                  error:(NSError **)error
-{
-    assert(storageItem);
-    NSData *itemData = [serializer serializeCredentialStorageItem:storageItem];
-    
-    if (!itemData)
-    {
-        if (error)
+        if (savedStorageItem)
         {
-            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Failed to serialize stored credential.", nil, nil, nil, context.correlationId, nil);
-            MSID_LOG_WITH_CTX(MSIDLogLevelError, context, @"Failed to serialize stored credential.");
+            [storageItem mergeStorageItem:savedStorageItem inBucket:bucket];
         }
         
-        return NO;
-    }
-    
-    MSID_LOG_WITH_CTX_PII(MSIDLogLevelInfo, context, @"Saving keychain item, item info %@.", MSID_PII_LOG_MASKABLE(storageItem));
-    
-    NSMutableDictionary *query = [self.defaultCacheQuery mutableCopy];
-    [query addEntriesFromDictionary:[self accountAttributeForKey:key]];
-    query[(id)kSecAttrService] = s_defaultKeychainLabel;
-    NSMutableDictionary *update = [NSMutableDictionary dictionary];
-    update[(id)kSecValueData] = itemData;
-    
-    __block OSStatus status;
-    dispatch_barrier_sync(s_synchronizationQueue, ^{
-        status = SecItemUpdate((CFDictionaryRef)query, (CFDictionaryRef)update);
-        MSID_LOG_WITH_CTX(MSIDLogLevelInfo, context, @"Keychain update status: %d.", (int)status);
-        
-        if (status == errSecItemNotFound)
-        {
-            [query addEntriesFromDictionary:update];
-            status = SecItemAdd((CFDictionaryRef)query, NULL);
-            MSID_LOG_WITH_CTX(MSIDLogLevelInfo, context, @"Keychain add status: %d.", (int)status);
-        }
-    });
-    
-    if (status != errSecSuccess)
-    {
-        MSID_LOG_WITH_CTX(MSIDLogLevelError, context, @"Failed to write item to keychain (status: %d).", (int)status);
-        [self createError:@"Failed to write item to keychain."
-                   domain:MSIDKeychainErrorDomain errorCode:status error:error context:context];
-        return NO;
+        [storageItem storeItem:credential inBucket:bucket forKey:key];
+        return [self saveStorageItem:storageItem key:key serializer:serializer context:context error:error];
     }
     
-    return YES;
+    return NO;
```

Comment:
log if unknown type found

Reply:
Removed the need to check key type in this layer as that logic is moved to the credential storage class.